Assets/MixedRealityToolkit/SpatialMapping/Scripts/TapToPlace.cs

```diff
@@ -149,35 +151,44 @@ public virtual void OnInputClicked(InputClickedEventData eventData)
         {
             // On each tap gesture, toggle whether the user is in placing mode.
             IsBeingPlaced = !IsBeingPlaced;
-            HandlePlacement();
+            HandlePlacement(eventData.InputSource);
             eventData.Use();
         }
 
-        private void HandlePlacement()
+        private void HandlePlacement(IInputSource inputSource)
         {
             if (IsBeingPlaced)
             {
-                StartPlacing();
+                StartPlacing(inputSource);
             }
             else
             {
-                StopPlacing();
+                StopPlacing(inputSource);
             }
         }
-        private void StartPlacing()
+
+        private void StartPlacing(IInputSource inputSource)
         {
             var layerCacheTarget = PlaceParentOnTap ? ParentGameObjectToPlace : gameObject;
             layerCacheTarget.SetLayerRecursively(IgnoreRaycastLayer, out layerCache);
             InputManager.Instance.PushModalInputHandler(gameObject);
 
+            // We have no input source if this call is coming from Start() because,
+            // the object was set to 'IsBeingPlaced'.
+            if (inputSource != null)
+            {
+                InputManager.Instance.RaisePlacingStarted(inputSource, 0);
+            }
+
             ToggleSpatialMesh();
             RemoveWorldAnchor();
         }
 
-        private void StopPlacing()
+        private void StopPlacing(IInputSource inputSource)
         {
             var layerCacheTarget = PlaceParentOnTap ? ParentGameObjectToPlace : gameObject;
             layerCacheTarget.ApplyLayerCacheRecursively(layerCache);
+            InputManager.Instance.RaisePlacingCompleted(inputSource, 0);
```

Comment:
@StephenHodgson I think this iteration addresses your concerns? 

My only concern with this version is, the PlacementEventData.selectedObject for OnPlacingCompleted is not the object that is being placed. For example, in the TapToPlace test scene the selectedObject is the 'roommesh' instead of the 'cube' that is being placed. That does not seem the desired behavior of these events?

Reply:
Yes, this is what I was.

I'll have to take a look at the class you're taking about, but the selectedObject is something that should be handled by the focus manager.

If you pop/pash the modal Input handler, you should be able to specify an object that will temp get all input directly.

Reply:
I believe the reason is:

In InputManager, even though the modalInput is 'cube' the the focusedObject is 'roomMesh' and the eventData.selectedObject is always set as the focusedObject.

Reply:
eventData.selectedObject should always be the focusedObject.

Reply:
So does it mean the behavior is by design and we should proceed without any modifications?

Summarizing the issue:
1. the selectedObject when OnPlacingStarted triggers is 'cube' as I would expect.
2. the selectedObject when OnPlacingCompleted triggers is 'roommesh' instead of the 'cube' that is being placed.

The events are routed to the listeners registered as expected. It's just the selectedObject when OnPlacingCompleted triggers that makes me wonder whether it's a bug or just by design. Ideally I would like to know what game object is being placed when those events fire ... but maybe I am overthinking the API.

Reply:
>Ideally I would like to know what game object is being placed when those events fire

You could add a GameObject property to the eventData to get this info.

Reply:
Oh that sounds good to me! I wasn't sure passing the GameObject around would be well received since there was no other event doing it,